### PR TITLE
Beast Client

### PR DESCRIPTION
**Pull request by jrbentzon**
## Scope

Implemented:
 - Beast Client `BeastClient` with basic submit and status check methods

## Checklist

- [ ] GitHub issue exists for this change.
- [ ] Unit tests added and they pass.
- [ ] Line Coverage is at least 80%.
- [x] Review requested on `latest` commit.

### CODE REVIEW

**Review by george-zubrienko — APPROVED** [review]

File: src/SnD.ApiClient/Beast/BeastClient.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+        do
+        {
+            await Task.Delay(pollInterval, cancellationToken);
+            result = await RequestState(requestId, cancellationToken);
+            if (completedStages.Contains(result.LifeCycleStage))
+            {
+                return result;
+            }
+        } while (!cancellationToken.IsCancellationRequested);
```
Comment (george-zubrienko) [inline src/SnD.ApiClient/Beast/BeastClient.cs:108]: should we instead throw when cancellation is requested inside the loop? that will ensure we return something or throw only

File: src/SnD.ApiClient/Beast/BeastClient.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        cancellationToken.ThrowIfCancellationRequested();
+        var requestUri = new Uri(baseUri, new Uri($"{apiVersion}/job/requests/tags/{clientTag}", UriKind.Relative));
+        var request = new HttpRequestMessage(HttpMethod.Get, requestUri);
+        var response = SendAuthenticatedRequestAsync(request, cancellationToken);
+        response.Result.EnsureSuccessStatusCode();
+        return JsonSerializer.Deserialize<string[]>(await response.Result.Content.ReadAsStringAsync(cancellationToken),
+            JsonSerializerOptions);
+    }
+
+    private async Task<RequestState> RequestState(string requestId, CancellationToken cancellationToken = default)
```
Comment (george-zubrienko) [inline src/SnD.ApiClient/Beast/BeastClient.cs:131]:
```suggestion
    private async Task<RequestState> GetRequestState(string requestId, CancellationToken cancellationToken = default)
```

File: src/SnD.ApiClient/Beast/BeastClient.cs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            }
+        } while (!cancellationToken.IsCancellationRequested);
+
+        return result;
+    }
+
+    /// <inheritdoc />
+    public Task<RequestState> GetJobStateAsync(string requestId, CancellationToken cancellationToken = default)
+    {
+        return RequestState(requestId, cancellationToken);
```
Comment (george-zubrienko) [inline src/SnD.ApiClient/Beast/BeastClient.cs:116]:
```suggestion
        return this.GetRequestState(requestId, cancellationToken);
```

File: src/SnD.ApiClient/Beast/BeastClient.cs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        return JsonSerializer.Deserialize<string[]>(await response.Result.Content.ReadAsStringAsync(cancellationToken),
+            JsonSerializerOptions);
+    }
+
+    private async Task<RequestState> RequestState(string requestId, CancellationToken cancellationToken = default)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+        var requestUri = new Uri(baseUri, new Uri($"{apiVersion}/job/requests/{requestId}", UriKind.Relative));
+        var request = new HttpRequestMessage(HttpMethod.Get, requestUri);
+        var response = SendAuthenticatedRequestAsync(request, cancellationToken);
```
Comment (george-zubrienko) [inline src/SnD.ApiClient/Beast/BeastClient.cs:136]:
```suggestion
        var response = await SendAuthenticatedRequestAsync(request, cancellationToken);
```

File: src/SnD.ApiClient/Beast/BeastClient.cs [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        response.Result.EnsureSuccessStatusCode();
```
Comment (george-zubrienko) [inline src/SnD.ApiClient/Beast/BeastClient.cs:137]:
```suggestion
        response.EnsureSuccessStatusCode();
```

File: src/SnD.ApiClient/Beast/BeastClient.cs [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        return JsonSerializer.Deserialize<RequestState>(
+            await response.Result.Content.ReadAsStringAsync(cancellationToken),
```
Comment (george-zubrienko) [inline src/SnD.ApiClient/Beast/BeastClient.cs:139]:
```suggestion
            await response.Content.ReadAsStringAsync(cancellationToken),
```

File: src/SnD.ApiClient/Crystal/CrystalClient.cs [hunk 7]
```diff
@@ -52,29 +64,68 @@ public class CrystalClient : SndApiClient, ICrystalClient
...
     /// <inheritdoc/>
-    public async Task<RunResult> GetResultAsync(string algorithm, string requestId, CancellationToken cancellationToken = default)
+    public async Task<CreateRunResponse> CreateRunAsync(string algorithm, JsonElement payload,
+        AlgorithmConfiguration customConfiguration, string tagId,
+        ConcurrencyStrategy concurrencyStrategy, CancellationToken cancellationToken)
     {
-        var requestUri = new Uri(baseUri, new Uri($"algorithm/{this.apiVersion}/results/{algorithm}/requests/{requestId}", UriKind.Relative));
+        if (string.IsNullOrEmpty(tagId))
+        {
+            throw new ArgumentException("TagId is required for concurrency strategy");
```
Comment (george-zubrienko) [inline src/SnD.ApiClient/Crystal/CrystalClient.cs:73]:
```suggestion
            throw new ArgumentException("You must supply a client tag when using a concurrency strategy");
```

File: src/SnD.ApiClient/Crystal/CrystalClient.cs [hunk 8]
```diff
@@ -108,23 +159,42 @@ public async Task<RunResult> AwaitRunAsync(string algorithm, string requestId, T
...
+
+
+    private async Task<RunResult[]> GetJobIdsByTagAsync(string algorithm, string clientTag,
+        CancellationToken cancellationToken = default)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+        var requestUri = new Uri(baseUri,
+            new Uri($"algorithm/{apiVersion}/results/{algorithm}/tags/{clientTag}", UriKind.Relative));
+        var request = new HttpRequestMessage(HttpMethod.Get, requestUri);
+        var response = SendAuthenticatedRequestAsync(request, cancellationToken);
```
Comment (george-zubrienko) [inline src/SnD.ApiClient/Crystal/CrystalClient.cs:194]:
```suggestion
        var response = await SendAuthenticatedRequestAsync(request, cancellationToken);
```
copilot much :)

File: src/SnD.ApiClient/Crystal/CrystalClient.cs [hunk 9]
```diff
@@ -108,23 +159,42 @@ public async Task<RunResult> AwaitRunAsync(string algorithm, string requestId, T
...
+        response.Result.EnsureSuccessStatusCode();
```
Comment (george-zubrienko) [inline src/SnD.ApiClient/Crystal/CrystalClient.cs:195]:
```suggestion
        response.EnsureSuccessStatusCode();
```

File: src/SnD.ApiClient/Crystal/CrystalClient.cs [hunk 10]
```diff
@@ -108,23 +159,42 @@ public async Task<RunResult> AwaitRunAsync(string algorithm, string requestId, T
...
+        return JsonSerializer.Deserialize<RunResult[]>(
+            await response.Result.Content.ReadAsStringAsync(cancellationToken),
```
Comment (george-zubrienko) [inline src/SnD.ApiClient/Crystal/CrystalClient.cs:197]:
```suggestion
            await response.Content.ReadAsStringAsync(cancellationToken),
```
Copilot greatness right here

File: src/SnD.ApiClient/Beast/BeastClient.cs (hunk 1 above)
Comment (jrbentzon) [reply src/SnD.ApiClient/Beast/BeastClient.cs:111]: I think the line above 
`cancellationToken.ThrowIfCancellationRequested();` will ensure that?

**Pull request merged**